Bluetooth: audio: ascs: Fix repeated ops->released callback call

MariuszSkamra opened the pull request:
Once the ACL disconnects, the implementation releases the related
endpoints. The ops->released callback is called once ASE enters IDLE
state, thus there is no need to call it explicitly.
This fixes repeated ops->released callback call.

MariuszSkamra committed:
tests: Bluetooth: ascs: Change bap_stream asserts to expectations

Continue test execution even if assumption failed. This won't break the
test execution prematurely and allow to perform a cleanup.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Fix repeated ops->released callback call

Once the ACL disconnects, the implementation releases the related
endpoints. The ops->released callback is called once ASE enters IDLE
state, thus there is no need to call it explicitly.
This fixes repeated ops->released callback call.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Add test for ASE release on ACL disconnection

This adds test case to verify if ACL disconnects, the associated ASE
is released.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Fix conn object being unreferenced prematurely

This fixes invalid conn unref that happens in ACL disconnected callback
context. When the ACL disconnects, the implementation waits for CIS to
be disconnected if connected. If the client won't disconnect the
CIS, the implementation initiates the disconnection by itself, from
delayed work context. Thus the ACL conn object has to be remained at
that time because it indicates whether the ASE object is in use or not.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

fredrikdanebjer commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -954,23 +953,6 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
-			}
-		}
-
-		if (stream != NULL && stream->conn != NULL) {
```
Why do you remove the unref here?

MariuszSkamra replied:
Because `bt_bap_stream_reset` called from `ascs_ep_set_state(BT_BAP_EP_STATE_IDLE)` does the same job.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -954,23 +953,6 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
 			ase_release(ase);
 			/* At this point, `ase` object have been free'd */
-
-			if (stream != NULL) {
-				const struct bt_bap_stream_ops *ops;
-
-				/* Notify upper layer */
-				ops = stream->ops;
-				if (ops != NULL && ops->released != NULL) {
-					ops->released(stream);
-				} else {
-					LOG_WRN("No callback for released set");
-				}
-			}
```
I might be missing something, but removing this may not be correct. 

As far as I can tell, the flow is

1. ACL Disconnects -> `disconnected`
2. `disconnected` -> `ase_release` if not `IDLE`
3. `ase_release` -> `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);`
4. `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);` -> `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);` if `BT_ISO_STATE_DISCONNECTED` else `ascs_disconnect_stream`

In other words, `ops->released` will only be called here if the CIS disconnects before the ACL (which is not a given). 

Your assumption (https://github.com/zephyrproject-rtos/zephyr/pull/57835/files#r1193347864) that `ase_release` will always call `ascs_ep_set_state(BT_BAP_EP_STATE_IDLE)` is incorrect

MariuszSkamra replied:
You're right, my comment (https://github.com/zephyrproject-rtos/zephyr/pull/57835/files#r1193347864) is a bit simplified, however the `ascs_ep_set_state(BT_BAP_EP_STATE_IDLE)` will be called eventually anyway.
Let me complete your flow:

> 1. ACL Disconnects -> `disconnected`
> 2. `disconnected` -> `ase_release` if not `IDLE`
> 3. `ase_release` -> `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);`
4a. if `BT_ISO_STATE_DISCONNECTED`: `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);`
4b. else `ascs_disconnect_stream`
5b. `ascs_iso_disconnected` -> `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);` if BT_BAP_EP_STATE_RELEASING

MariuszSkamra replied:
@Thalley I've added unit test for this particular case `test_release_ase_on_acl_disconnection_server_terminates_cis`

Thalley replied:
> ascs_ep_set_state(BT_BAP_EP_STATE_IDLE) will be called eventually anyway.

Agreed, but the difference is whether or not the stream object is reusable right after the application gets the ACL disconnect callback. 

The code to release the stream here was added for that purpose, and this would have broken then shell earlier as it expected the streams to be released at ACL disconnect. 

Now that I think about it, since we do not know the order of the app vs ASCS ACL disconnect callback, the application can still not expect the streams to be released by the time the callback gets called.

I am OK with this change, and I see that the shell no longer triggers the unicast group delete on ACL disconnect, but rather on stream released.

Thalley commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -220,3 +220,79 @@ ZTEST_F(ascs_test_suite, test_abort_client_operation_if_callback_not_registered)
+	struct bt_iso_chan *chan;
+	uint8_t ase_id;
+
+	if (IS_ENABLED(CONFIG_BT_ASCS_ASE_SNK)) {
+		ase = fixture->ase_snk.attr;
+		ase_id = fixture->ase_snk.id;
+	} else {
+		ase = fixture->ase_src.attr;
+		ase_id = fixture->ase_src.id;
+	}
+
+	zexpect_not_null(ase);
+	zexpect_true(ase_id != 0x00);
+
+	bt_bap_unicast_server_register_cb(&mock_bap_unicast_server_cb);
+
+	/* Set ASE to non-idle state */
+	test_preamble_state_streaming(conn, ase_id, stream, &chan,
+				      !IS_ENABLED(CONFIG_BT_ASCS_ASE_SNK));
```
Instead of relying on `IS_ENABLED(CONFIG_BT_ASCS_ASE_SNK)` would it make more sense to test this with
1) Sink stream in CIS only
2) Source stream in CIS only
3) Sink and Source stream in same CIS
?

Thalley commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -220,3 +220,79 @@ ZTEST_F(ascs_test_suite, test_abort_client_operation_if_callback_not_registered)
+	test_preamble_state_streaming(conn, ase_id, stream, &chan,
+				      !IS_ENABLED(CONFIG_BT_ASCS_ASE_SNK));
+
+	/* Mock ACL disconnection */
+	mock_bt_conn_disconnected(conn);
+
+	/* Server waits for CIS to be disconnected by client */
+	k_sleep(K_MSEC(CONFIG_BT_ASCS_ISO_DISCONNECT_DELAY));
+
+	/* Expected to notify the upper layers */
+	expect_bt_bap_unicast_server_cb_release_called_once(stream);
+	expect_bt_bap_stream_ops_released_called_once(stream);
```
The `CONFIG_BT_ASCS_ISO_DISCONNECT_DELAY` is only used to disconnect the CIS if the stream goes out of the streaming state; it is not related to CIS disconnect events due to ACL disconnects. 

I don't see the value of this specific test case, as neither the client nor the server initiates the CIS disconnect if the ACL disconnects

MariuszSkamra replied:
Cannot agree, as ACL disconnection triggers the `ascs_disconnect_stream` if CIS has been not disconnected before ACL (which is possible case - allowed by the spec).

Thalley replied:
Hmm, I see your point, however there is not really any value in attempting to disconnect the CIS _after_ the ACL disconnect event: We know the CIS will get disconnected right after (if not before, depending on the controller implementation). 

IMO the host should not attempt to disconnect the CIS on ACL disconnect, which I can see that we do in ASCS - Maybe that part needs changed? There no value in calling `ascs_disconnect_stream` if ACL state is not `BT_CONN_STATE_CONNECTED`.

Thalley commented on tests/bluetooth/audio/mocks/src/conn.c:
```diff
@@ -29,3 +29,12 @@ void bt_conn_unref(struct bt_conn *conn)
 
 }
+
+void mock_bt_conn_disconnected(struct bt_conn *conn)
+{
+	STRUCT_SECTION_FOREACH(bt_conn_cb, cb) {
+		if (cb->disconnected) {
+			cb->disconnected(conn, 0x13);
+		}
+	}
+}
```
Consider add the reason as a parameter here; we may want to test with both `BT_HCI_ERR_REMOTE_USER_TERM_CONN` and `BT_HCI_ERR_CONN_TIMEOUT`

Thalley commented on tests/bluetooth/audio/mocks/src/kernel.c:
```diff
@@ -54,6 +56,28 @@ int k_work_cancel_delayable(struct k_work_delayable *dwork)
+		if (work->flags & K_WORK_DELAYED) {
+			struct k_work_delayable *dwork;
+
+			dwork = CONTAINER_OF(work, struct k_work_delayable, work);
```
Not a change needed, but could consider using `k_work_delayable_from_work`

MariuszSkamra committed:
tests: Bluetooth: ascs: Add test for stream pair release on ACL disc

This adds test case to verify if ACL disconnects, the associated
stream pair is released.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
Added 2 more test cases `test_release_stream_pair_on_acl_disconnection_client_terminates_cis` and `test_release_stream_pair_on_acl_disconnection_server_terminates_cis` to verify whether stream pair is released on ACL disconnection

Thalley commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -220,3 +221,203 @@ ZTEST_F(ascs_test_suite, test_abort_client_operation_if_callback_not_registered)
+
+	bt_bap_unicast_server_register_cb(&mock_bap_unicast_server_cb);
+
+	test_ase_control_client_config_codec(conn, ase_snk_id, &snk_stream);
+	test_ase_control_client_config_qos(conn, ase_snk_id);
+	test_ase_control_client_enable(conn, ase_snk_id);
+
+	test_ase_control_client_config_codec(conn, ase_src_id, &src_stream);
+	test_ase_control_client_config_qos(conn, ase_src_id);
+	test_ase_control_client_enable(conn, ase_src_id);
+
+	err = mock_bt_iso_accept(conn, 0x01, 0x01, &chan);
+	zassert_equal(0, err, "Failed to connect iso: err %d", err);
+
+	test_ase_control_client_receiver_start_ready(conn, ase_src_id);
+
+	err = bt_bap_stream_start(&snk_stream);
+	zassert_equal(0, err, "bt_bap_stream_start err %d", err);
```
Is it not possible to use `test_preamble_state_streaming` here?

MariuszSkamra replied:
This requires refactor as well, I'll create an issue instead

Thalley commented on tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:
```diff
@@ -155,6 +155,26 @@ do {
 				       func_name, "stream");))                                     \
 } while (0)
 
+#define expect_bt_bap_unicast_server_cb_release_called_twice(_streams)                             \
```
This seems like it'll scale poorly :D 

Does it make sense to have a `expect_bt_bap_unicast_server_cb_release_called(_cnt, _streams)`? as a more generic approach? 
Not just for this, for for all the `expect_..._called_...`?

Thalley commented on tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:
```diff
@@ -155,6 +155,26 @@ do {
+	const char *func_name = "bt_bap_unicast_server_cb.release";                                \
+												   \
+	zassert_equal(2, mock_bap_unicast_server_cb_release_fake.call_count,                       \
+		      "'%s()' was called %u times, but expected once",                             \
```
```suggestion
		      "'%s()' was called %u times, but expected twice",                             \
```

MariuszSkamra committed:
tests: Bluetooth: ascs: Factor out call count verification function

Reduce code duplicates by using common function to verify mock function
call count.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

szymon-czapracki commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -220,3 +221,203 @@ ZTEST_F(ascs_test_suite, test_abort_client_operation_if_callback_not_registered)
+
+ZTEST_F(ascs_test_suite, test_release_ase_on_acl_disconnection_client_terminates_cis)
+{
+	const struct bt_gatt_attr *ase;
+	struct bt_bap_stream *stream = &fixture->stream;
+	struct bt_conn *conn = &fixture->conn;
+	struct bt_iso_chan *chan;
+	uint8_t ase_id;
```
Keep definitions from longest to shortest for code style conformity.
Applies in multiple places.

MariuszSkamra replied:
done
